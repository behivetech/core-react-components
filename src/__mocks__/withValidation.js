import {omit} from 'lodash';

export default (WrappedComponent) => function FormFieldValidation(props) { 
    return <WrappedComponent {...omit(props, ['formState'])} />; 
};