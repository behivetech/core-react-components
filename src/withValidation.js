// Vendor Libs
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {
    debounce,
    isFunction,
    transform,
} from 'lodash';

// Core Libs
import * as validators from './lib-core/validators';

export default function (WrappedComponent) {
    class FormFieldValidation extends Component {
        constructor(props) {
            super(props);
            this.handleChange = this.handleChange.bind(this);
            this.validateFieldDebounced = debounce(this.validateField, 200, {trailing: true});
            this.state = {fieldError: null};
        }

        componentDidMount() {
            this.validateField(this.props.value);    
        }

        validateField(value) {
            const {formState, name} = this.props;
            let fieldError = null;

            // Using transform to iterate through the validators until an error is found.
            // Only sending one error at a time to the form field to prevent a mass
            // amount of errors showing up. This is also a much more efficient approach
            // so all validators aren't iterated through every time the field needs to
            // be validated.
            transform(this.props.validate, (errors, validator) => {
                if (isFunction(validator)) {
                    fieldError = validator(value, formState.getFieldValues());
                } else {
                    const importedValidator = validators[validator]; // eslint-disable-line

                    if (importedValidator && isFunction(importedValidator)) {
                        fieldError = importedValidator(value);
                    } else {
                        fieldError = `${validator} is not a valid validator.`;
                    }
                }

                return !fieldError;
            });

            (fieldError) ? formState.setFieldError(name) : formState.unsetFieldError(name);
            this.setState({fieldError});
        }

        handleChange(event) {
            this.validateFieldDebounced(event.target.value);
            this.props.onChange(event);
        }

        getError() {
            return (this.props.formState.getInitialSubmit()) ? this.state.fieldError : null;
        }

        render() {
            return (
                <WrappedComponent
                    {...this.props}
                    error={this.getError()}
                    onChange={this.handleChange}
                />
            );
        }
    }

    FormFieldValidation.propTypes = {
        validate: PropTypes.arrayOf(PropTypes.oneOfType(
            [PropTypes.string, PropTypes.func]
        )).isRequired,
    };

    return FormFieldValidation;
}

