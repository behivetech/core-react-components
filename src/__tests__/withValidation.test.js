import withValidation from '../withValidation';
// Needed for validateFieldDebounced
jest.useFakeTimers();

describe('withValidation', () => {
    const MockComponentWrapped = (props) => <div />;
    const MockComponentControlled = withValidation(MockComponentWrapped);
    // eslint-disable-next-line
    const MockComponent = (props) => <MockComponentControlled {...props} />;
    const getProps = (additionalProps = {}, initialSubmit = false) => {
        return {
            className: 'mock-class',
            formState: {
                getFieldValues: jest.fn(() => {
                    return {mockField: 'mock value'};
                }),
                getInitialSubmit: jest.fn(() => initialSubmit),
                setFieldError: jest.fn(),
                unsetFieldError: jest.fn(),
            },
            name: 'mockName',
            onChange: jest.fn(),
            validate: [jest.fn(() => 'mock error')],
            ...additionalProps,
        };
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render component and run validations on mount', () => {
        const mockProps = getProps();
        const wrapper = mount(<MockComponent {...mockProps} value="mock value" />);

        expect(wrapper).toMatchSnapshot();
        expect(mockProps.validate[0]).toBeCalledWith('mock value', {mockField: 'mock value'})
    });

    it('should validate with an imported validation function and error should be null', () => {
        const mockProps = getProps({validate: ['required']});
        const wrapper = shallow(<MockComponent {...mockProps} />).dive();

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.state().fieldError).toBe('This field cannot be empty');
        expect(wrapper.props().error).toBe(null);
        expect(mockProps.formState.setFieldError).toBeCalledWith(mockProps.name);
    });

    it('should return an error if the imported validation function does not exist', () => {
        const mockProps = getProps({validate: ['mockValidation']});
        const wrapper = shallow(<MockComponent {...mockProps} />).dive();

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.state().fieldError).toBe('mockValidation is not a valid validator.');
    });

    it('should pass an error prop to the component when initialSubmit is true', () => {
        const mockProps = getProps({}, true);
        const wrapper = shallow(<MockComponent {...mockProps} />).dive();

        expect(wrapper.props().error).toBe('mock error');
    });

    it('should handle a change correctly and unset a field error', () => {
        const mockProps = getProps({validate: ['required']});
        const mockEvent = {target: {value: 'mock value 2'}};
        const wrapper = shallow(<MockComponent {...mockProps} />).dive();
        const wrapperInstance = wrapper.instance();

        wrapperInstance.handleChange(mockEvent);
        // Waiting for validateFieldDebounced to run
        jest.runAllTimers();
        expect(wrapperInstance.props.onChange).toBeCalledWith(mockEvent);
        expect(wrapperInstance.props.formState.unsetFieldError).toBeCalledWith(mockProps.name);
    });
});