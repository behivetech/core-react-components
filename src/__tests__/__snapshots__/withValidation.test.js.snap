// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withValidation should render component and run validations on mount 1`] = `
<MockComponent
  className="mock-class"
  formState={
    Object {
      "getFieldValues": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Object {
              "mockField": "mock value",
            },
          },
        ],
      },
      "getInitialSubmit": [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": false,
          },
          Object {
            "isThrow": false,
            "value": false,
          },
        ],
      },
      "setFieldError": [MockFunction] {
        "calls": Array [
          Array [
            "mockName",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
      "unsetFieldError": [MockFunction],
    }
  }
  name="mockName"
  onChange={[MockFunction]}
  validate={
    Array [
      [MockFunction] {
        "calls": Array [
          Array [
            "mock value",
            Object {
              "mockField": "mock value",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": "mock error",
          },
        ],
      },
    ]
  }
  value="mock value"
>
  <FormFieldValidation
    className="mock-class"
    formState={
      Object {
        "getFieldValues": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": Object {
                "mockField": "mock value",
              },
            },
          ],
        },
        "getInitialSubmit": [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": false,
            },
            Object {
              "isThrow": false,
              "value": false,
            },
          ],
        },
        "setFieldError": [MockFunction] {
          "calls": Array [
            Array [
              "mockName",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "unsetFieldError": [MockFunction],
      }
    }
    name="mockName"
    onChange={[MockFunction]}
    validate={
      Array [
        [MockFunction] {
          "calls": Array [
            Array [
              "mock value",
              Object {
                "mockField": "mock value",
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": "mock error",
            },
          ],
        },
      ]
    }
    value="mock value"
  >
    <MockComponentWrapped
      className="mock-class"
      error={null}
      formState={
        Object {
          "getFieldValues": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": Object {
                  "mockField": "mock value",
                },
              },
            ],
          },
          "getInitialSubmit": [MockFunction] {
            "calls": Array [
              Array [],
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": false,
              },
              Object {
                "isThrow": false,
                "value": false,
              },
            ],
          },
          "setFieldError": [MockFunction] {
            "calls": Array [
              Array [
                "mockName",
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          },
          "unsetFieldError": [MockFunction],
        }
      }
      name="mockName"
      onChange={[Function]}
      validate={
        Array [
          [MockFunction] {
            "calls": Array [
              Array [
                "mock value",
                Object {
                  "mockField": "mock value",
                },
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": "mock error",
              },
            ],
          },
        ]
      }
      value="mock value"
    >
      <div />
    </MockComponentWrapped>
  </FormFieldValidation>
</MockComponent>
`;

exports[`withValidation should return an error if the imported validation function does not exist 1`] = `
<MockComponentWrapped
  className="mock-class"
  error={null}
  formState={
    Object {
      "getFieldValues": [MockFunction],
      "getInitialSubmit": [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": false,
          },
          Object {
            "isThrow": false,
            "value": false,
          },
        ],
      },
      "setFieldError": [MockFunction] {
        "calls": Array [
          Array [
            "mockName",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
      "unsetFieldError": [MockFunction],
    }
  }
  name="mockName"
  onChange={[Function]}
  validate={
    Array [
      "mockValidation",
    ]
  }
/>
`;

exports[`withValidation should validate with an imported validation function and error should be null 1`] = `
<MockComponentWrapped
  className="mock-class"
  error={null}
  formState={
    Object {
      "getFieldValues": [MockFunction],
      "getInitialSubmit": [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": false,
          },
          Object {
            "isThrow": false,
            "value": false,
          },
        ],
      },
      "setFieldError": [MockFunction] {
        "calls": Array [
          Array [
            "mockName",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
      "unsetFieldError": [MockFunction],
    }
  }
  name="mockName"
  onChange={[Function]}
  validate={
    Array [
      "required",
    ]
  }
/>
`;
